src/core/algorithms/sfd/sfd_algorithm.cpp

```diff
@@ -0,0 +1,302 @@
+            if (n_i[i] * n_j[j] == 0) return 0;
+            long double actual = n_i_j[i][j];
+            long double expected = n_i[i] * n_j[j] / sample_size;
+            chi_squared += (actual - expected) * (actual - expected) / (expected);
+        }
+    }
+    return chi_squared;
+}
+
+void SFDAlgorithm::Init(model::ColumnIndex columns) {
+    is_skewed_.resize(columns, false);
+    domains_.resize(columns, 0);
+}
+
+std::pair<model::ColumnIndex, model::ColumnIndex> SFDAlgorithm::SwitchIndicesIfNecessary(
+        model::ColumnIndex ind1, model::ColumnIndex ind2, size_t card1, size_t card2) {
+    if (card2 > card1) {
+        return {ind2, ind1};
+    }
+    return {ind1, ind2};
+}
+
+bool SFDAlgorithm::ChiSquaredTest(model::ColumnIndex col_i, model::ColumnIndex col_k, Sample &smp,
+                                  std::vector<std::vector<long double>> const &n_i_j,
+                                  std::vector<long double> const &n_i,
+                                  std::vector<long double> const &n_j) const {
+    long double chi_squared =
+            CalculateChiSquared(n_i_j, n_i, n_j, col_i, col_k, smp.GetRowIndices().size());
+
+    long double v = (domains_[col_i] - 1) * (domains_[col_k] - 1);
+
+    boost::math::chi_squared dist(v);
+
+    long double t = quantile(dist, 1 - max_false_positive_probability_);
+    return chi_squared > t;
+}
+
+void SFDAlgorithm::CheckForCorrelation(model::ColumnIndex col_i, model::ColumnIndex col_k,
+                                       std::vector<model::TypedColumnData> const &data, Sample &smp,
+                                       FrequencyHandler const &handler) {
+    SkewHandling(col_i, col_k, data, smp, handler);
+
+    std::vector<std::vector<long double>> n_i_j(domains_[col_i],
+                                                std::vector<long double>(domains_[col_k], 0));
+    std::vector<long double> n_i(domains_[col_i], 0);
+    std::vector<long double> n_j(domains_[col_k], 0);
+
+    FillContingencyTable(smp, data, col_i, col_k, n_i_j, n_i, n_j, handler);
+
+    if (TooMuchStructuralZeros(n_i_j, col_i, col_k) ||
+        ChiSquaredTest(col_i, col_k, smp, n_i_j, n_i, n_j)) {
+        correlations_.emplace_back(col_i, col_k);
+        return;
+    }
+    unrelated_.emplace_back(col_i, col_k);
+}
+
+bool SFDAlgorithm::IsTrivialCase(model::ColumnIndex col_ind, std::vector<bool> &is_soft_or_trivial,
+                                 FrequencyHandler const &handler, size_t row_count) {
+    if (is_soft_or_trivial[col_ind]) {
+        return true;
+    }
+    if (handler.GetColumnCardinality(col_ind) >= (1 - minimum_cardinality_) * row_count) {
+        soft_keys_.push_back(col_ind);
+        is_soft_or_trivial[col_ind] = true;
+        return true;
+    }
+    if (handler.GetColumnCardinality(col_ind) == 1) {
+        trivial_columns_.push_back(col_ind);
+        is_soft_or_trivial[col_ind] = true;
+        return true;
+    }
```

Comment:
From this code it follows that the soft column cannot be trivial. Do we really want this?

Reply:
According to the pseudo-code from the original paper soft column cannot be trivial.